### PR TITLE
Kafka "Hello World"

### PR DESCRIPTION
**Pull request by LucyxxNie**

### CODE REVIEW

File: src/kafka/consumer.clj [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+
+
+(comment
+  (receive-msg-single-thread consumer-properties)
+  #_=> Message: Hello World
+
+
+  (receive-msg-multi-thread consumer-properties)
+  #_=> Message: Hello World
```
Comment (lukaszhall) [inline src/kafka/consumer.clj:60]: This is not the value produced by this fn

File: src/kafka/consumer.clj [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    (org.apache.kafka.common.errors WakeupException)))
+
+(def consumer-properties
+  {"client.id",         (.getHostName (InetAddress/getLocalHost))
+   "group.id",          "group1"
+   "bootstrap.servers", "localhost:9092"
+   "key.deserializer"   StringDeserializer
+   "value.deserializer" StringDeserializer})
+
+(defn consumer-subscribe
```
Comment (lukaszhall) [inline src/kafka/consumer.clj:16]: dangerous or side-effect fns should be named as such. e.g., `!`

File: src/kafka/consumer.clj [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+(defn receive-msg-single-thread
+  [consumer-properties]
+  (let [consumer (KafkaConsumer. consumer-properties)
+        topic    "example-topic"]
+    (consumer-subscribe consumer topic)
+    (while true
+      (let [records (.poll consumer (Duration/ofMillis 1))]
+        (doseq [record records]
+          (println "Message: " (.value record)))))))
```
Comment (lukaszhall) [inline src/kafka/consumer.clj:29]: Only need 1 fn to demonstrate receiving messages. That fn should follow basic data-oriented principles. 
* Why does this fn block?
* Why does this fn not return data?

Comment (lukaszhall) [reply src/kafka/consumer.clj:29]: _Poor_ decomposition. Why is everything being slammed together in one function? For example, how is building a consumer related to polling for records?

File: src/kafka/consumer.clj [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+
+(comment
+  (receive-msg-single-thread consumer-properties)
+  #_=> Message: Hello World
+
+
+  (receive-msg-multi-thread consumer-properties)
+  #_=> Message: Hello World
+  )
```
Comment (lukaszhall) [inline src/kafka/consumer.clj:61]: This RCB is missing most of what we would expect from an RCB. 

Also, combine your namespaces. The current layout is not wrong, but you'll want them combined for ease of demonstration of the next concepts

Comment (LucyxxNie) [reply src/kafka/consumer.clj:61]: Do you want the producer and consumer in the same ns?

Comment (lukaszhall) [reply src/kafka/consumer.clj:61]: yes

File: src/kafka/producer_consumer.clj [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  (:import
+    (java.time Duration)
+    (java.net InetAddress)
+    (org.apache.kafka.common.errors WakeupException)
+    (org.apache.kafka.common.serialization StringSerializer StringDeserializer)
+    (org.apache.kafka.clients.producer KafkaProducer ProducerRecord)
+    (org.apache.kafka.clients.consumer KafkaConsumer)))
+
+(def producer-properties
+  {"client.id",         (.getHostName (InetAddress/getLocalHost))
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: keep this a just data - no fn calls

File: src/kafka/producer_consumer.clj [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+  [producer-properties]
+  (KafkaProducer. producer-properties))
+
+(defn build-consumer
+  [consumer-properties]
+  (KafkaConsumer. consumer-properties))
+
+(defn new-msg
+  [topic key msg]
+  (println "Msg created successfully.")
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: Comment out or remove debug code. (See below as well)

File: src/kafka/producer_consumer.clj [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+  (ProducerRecord. topic key msg))
+
+(defn send-msg!
+  [producer msg-value]
+  (let [producer-topic "example-topic"
+        msg            (new-msg producer-topic "msg" msg-value)]
+    (println "Msg sent successfully!")
+    (.send producer msg)
+    (.close producer)))
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: You should have only a single return value in your `let` forms. Everything else should be in the `binding` blocks (for readability). e.g,
```
(let [a 1
      b ,,,,,
      c ,,,,,,]
    c)

File: src/kafka/producer_consumer.clj [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+
+(defn consumer-subscribe!
+  [consumer topic]
+  (.subscribe consumer [topic])
+  (.commitAsync consumer))
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: Problem decomposition issue -- Isn't subscribing considered part of setting up a consumer?
Also, `commit` is only necessary for updating offsets (i.e., after reading messageS)

File: src/kafka/producer_consumer.clj [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+
+(defn receive-msg-single-thread
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: Unnecessary naming additions. This also has side effects and is not idempotent, it should be name as such!

File: src/kafka/producer_consumer.clj [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+  [consumer topic]
+  (consumer-subscribe! consumer topic)
+  (let [records (.poll consumer (Duration/ofMillis 100000))]
+    (mapv #(str "Msg: " (.value %)) records)))
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: 1. Let bindings should have the return value as their last element. Keep other code in the `[]`
2. Why is the msg value being polluted with additional strings?
3. Use threading here
4. avoid # function forms (for now)

File: src/kafka/producer_consumer.clj [hunk 11]
```diff
@@ -0,0 +1,119 @@
...
+          (throw e)))
+      (finally (.close consumer)))))
+
+(defn shutdown
+  [state consumer]
+  (reset! state true)
+  (.wake consumer))
+
+(comment
+  (do producer-properties)
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: why `do`? Why not just the values?

File: src/kafka/producer_consumer.clj [hunk 12]
```diff
@@ -0,0 +1,119 @@
...
+  #_=> #object[org.apache.kafka.clients.producer.KafkaProducer
+               0x6e7e1683
+               "org.apache.kafka.clients.producer.KafkaProducer@6e7e1683"]
+
+  (build-consumer consumer-properties)
+  #_=> #object[org.apache.kafka.clients.consumer.KafkaConsumer
+               0x476b779f
+               "org.apache.kafka.clients.consumer.KafkaConsumer@476b779f"]
+
+  (new-msg "topic" "key" "value-msg")
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj:101]: Why would someone using this namespace need to create kafka ProducerRecords?

File: src/kafka/producer_consumer.clj [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+  (KafkaConsumer. consumer-properties))
+
+(defn new-msg
+  [topic key msg]
+  (println "Msg created successfully.")
+  (ProducerRecord. topic key msg))
+
+(defn send-msg!
+  [producer msg-value]
+  (let [producer-topic "example-topic"
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: isn't this considered producer configuration?

File: src/kafka/producer_consumer.clj [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+               "ProducerRecord(topic=topic, partition=null, headers=RecordHeaders(headers = [], isReadOnly = false), key=key, value=value-msg, timestamp=null)"]
+
+  (send-msg! (build-producer producer-properties) "Message content")
+  ;Msg created successfully.
+  ;Msg sent successfully!
+  #_=> nil
+
+  (consumer-subscribe! (build-consumer consumer-properties)
+                       "Parse the topic consumer wants to listen")
+  #_=> nil
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: Why isn't this abstracted as part of build?

File: src/kafka/producer_consumer.clj [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        state (atom false)]
+    (try
+      (consumer-subscribe! consumer topic)
+      (while (not @state)
+        (let [records (.poll consumer (Duration/ofMillis 1))]
+          (mapv #(str "Message: " (.value %)) records)))
+      (catch WakeupException e
+        (when-not @state
+          (throw e)))
+      (finally (.close consumer)))))
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: Remove this fn. Demonstrate simple usage patterns in RCB/repl.

File: src/kafka/producer_consumer.clj [hunk 16]
```diff
@@ -0,0 +1,119 @@
...
+
+(defn shutdown
+  [state consumer]
+  (reset! state true)
+  (.wake consumer))
```
Comment (lukaszhall) [inline src/kafka/producer_consumer.clj]: Not sure why this fn is here/exists?

**Pull request merged**